File: src/algorithms/binary-trees/invert-binary-tree/solution1.go
package main

type BinaryTree struct {
	Value int

	Left *BinaryTree
	Right *BinaryTree
}

// O(n) time | O(n) space
func (tree *BinaryTree) InvertBinaryTree() {
	queue := []*BinaryTree{tree}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		if current == nil {
			continue
		}
		current.Left, current.Right = current.Right, current.Left
		queue = append(queue, current.Left, current.Right)
	}
}

File: src/algorithms/binary-trees/invert-binary-tree/solution1.ts
class BinaryTree {
  value: number;
  left: BinaryTree | null;
  right: BinaryTree | null;

  constructor(value: number) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

// Iterative Solution
// O(n) time | O(n) space
export function invertBinaryTree(tree: BinaryTree | null) {
  const queue: Array<BinaryTree | null> = [tree];
  while (queue.length) {
  	const current = queue.shift()!;
  	if (current === null) continue;
  	swapLeftAndRight(current);
  	queue.push(current.left);
  	queue.push(current.right);
  };
};

function swapLeftAndRight(tree: BinaryTree) {
	const left = tree.left;
	tree.left = tree.right;
	tree.right = left
};
